### ISSUE TITLE
Multiple Next apps on single domain will collide

### ISSUE THREAD

**Issue by seldo** [zeit/next.js#257]
Next uses routes like `/_next` and `/static` by default. If you wanted to run multiple different Next apps on the same domain, they would collide with each other. Can these paths be made configurable?

**Comment by rauchg** [comment]
I believe I opened an issue for _mounting_, or I should have otherwise. The idea is that you can supply a prefix in the next config (like `/some/sub/app`), and then you can merge next apps at the proxy level.

**Comment by rauchg** [comment]
There are a few caveats however. For example, `<img src="/static/logo.png">` won't work. Maybe we should introduce `this.context.next.prefix` or something like that.

cc @nkzawa

**Comment by seldo** [comment]
Configurable prefixes would work perfectly for our use case.

**Comment by nkzawa** [comment]
The problem of `this.context.next.prefix` is the value goes wrong when you load components from other app using `<Link/>`.

For example, when there are two apps which mounted to `/a` and `/b`.
`<img src={this.context.next.prefix + '/static/logo.png'} />` becomes `<img src="/a/static/logo.png" />` or `<img src="/b/static/logo.png" />` depends on which mounted app the component is loaded.

**Comment by rauchg** [comment]
Another option is a compile-time thing.

```
import { pathPrefix } from 'next/env'
```

and we can replace it with webpack

**Comment by nkzawa** [comment]
Another options is to fetch the value on `.json` request.

`/b/page.json?pathPrefix=1`

```js
// response
{
  "component": "...",
  "pathPrefix": "/b"
}
```

Maybe we can detect if it's an external component by comparing to own `pathPrefix` and automatically add the query parameter to the request.

**Comment by cncolder** [comment]
Cannot waiting this option.

Now I extends `next/dist/server` and override `defineRoutes()`.
extends `NextScript` in `next/document` and override `render()`.

Client side I monkey path `XMLHttpRequest.prototype.open`

**Comment by rauchg** [comment]
We won't be implementing it in the short term. If anyone wants to work on it, we could certainly provide feedback on suggested designs

**Comment by albinekb** [comment]
@rauchg 
>There are a few caveats however. For example, <img src="/static/logo.png"> won't work. Maybe we should introduce this.context.next.prefix or something like that.

Maybe by using the `<Base>`tag? 🤔 
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base

**Comment by JohnPhoto** [comment]
@cncolder Do you have a gist or something illustrating your overrides?
We need this functionality at our site as well.

**Comment by ianatha** [comment]
@JohnPhoto, these are the monkey-patches we've done to support this.

In `_document.js`:

```
import Document, { Head, Main, NextScript } from 'next/document'
import htmlescape from 'htmlescape'
import Router from 'next/router'

class PrefixedNextScript extends NextScript {
  render () {
    const { staticMarkup, __NEXT_DATA__ } = this.context._documentProps
    let { buildId } = __NEXT_DATA__

    return <div>
      {staticMarkup ? null : <script dangerouslySetInnerHTML={{
        __html: `__NEXT_DATA__ = ${htmlescape(__NEXT_DATA__)}; module={};`
      }} />}
    <script dangerouslySetInnerHTML={{
            __html:
`
__NEXT_DATA__.prefix = "` + this.props.prefix + `";
var open = XMLHttpRequest.prototype.open;

XMLHttpRequest.prototype.open = function() {
  if (arguments[1].indexOf("/_next/") == 0 || arguments[1].indexOf("/__webpack_hmr") == 0) {
    arguments[1] = __NEXT_DATA__.prefix + arguments[1];
  }
  open.apply(this, arguments);
};
` }} />
      { staticMarkup ? null : <script type='text/javascript' src={this.props.prefix + `/_next/${buildId}/commons.js`} /> }
      { staticMarkup ? null : <script type='text/javascript' src={this.props.prefix + `/_next/${buildId}/main.js`} /> }
    </div>
  }
}

const ROOT_URL = (process.env.ROOT_URL ? process.env.ROOT_URL : "");

export default class MyDocument extends Document {
  render () {
    return (
      <html>
        <Head>
          <title>Title Here</title>
          <link rel="stylesheet" type="text/css" href={ROOT_URL + "/static/main.css"} />
        </Head>
        <body>
          <Main />
          <PrefixedNextScript prefix={ROOT_URL} />
        </body>
      </html>
    )
  }
}

Router.ready(() => {
  if (!Router.router.change_monkeypatched) {
    let router_change = Router.router.change;
    Router.router.change = function(a, b, c) {
      arguments[1] = "" + arguments[1]
      arguments[2] = __NEXT_DATA__.prefix + arguments[2]
      return router_change.apply(this, arguments);
    }
    Router.router.change_monkeypatched = true;
  }
});
```

Hope this helps.

**Comment by nndung179** [comment]
Above solution is now working in version 2.1.1

Here is the replacement:

```js
class PrefixedNextScript extends NextScript {
  render() {
    const { __NEXT_DATA__ } = this.context._documentProps;
    const { prefix } = this.props;

    const env = process.env.NODE_ENV;

    // script element contain development js
    let scriptElements = (
      <div>
        <script type="text/javascript" src={`${prefix}/_next/-/manifest.js`} />
        <script type="text/javascript" src={`${prefix}/_next/-/commons.js`} />
        <script type="text/javascript" src={`${prefix}/_next/-/main.js`} />
      </div>
    );

    if (env === 'production') {
      const { buildStats } = __NEXT_DATA__;
      const app = 'app.js';

      // script element contain production js
      scriptElements =
        <script
          type="text/javascript"
          src={`${prefix}/_next/${buildStats[app].hash}/${app}`}
        />
    }

    return (
      <div>
        <script
          dangerouslySetInnerHTML={
            { __html: `__NEXT_DATA__ = ${htmlescape(__NEXT_DATA__)}; module={};` }
          }
        />
        {scriptElements}
      </div>
    );
  }
}

```

**Comment by tpai** [comment]
This issue should be close, `2.2.0` release `assetPrefix` config flag.

**Comment by arunoda** [comment]
@tpai does the assetPrefix fix this issue. 
I like to learn more about it. 
(Anyway, this was not our intention, it this fixes that would be awesome)

**Comment by nndung179** [comment]
Hi @arunoda.

`assetPrefix` does not fix the issues. It only give a prefix to resources in static dir. If it can support bundle js. If so is going to be great. 

For example, when run command to build in production mode. The `assetPrefix` does not include the prefix to that script.

**Comment by tpai** [comment]
@nndung179 `assetPrefix` does support bundle js, you could check this [server/document.js:46](https://github.com/zeit/next.js/blob/v3-beta/server/document.js#L46), and [REAMDE](https://github.com/zeit/next.js#cdn-support-with-asset-prefix) also mentioned.

@arunoda Not perfectly fixed, but it still work.

next.config.js

```javascript
module.exports = {
  assetPrefix: '/prefix/'
};
```
customServer.js

```javascript
express.use((req, res) => {
  req.url = req.url.replace('/prefix', '');
  handle(req, res);
});
```

**Comment by tscanlin** [comment]
I had a PR to fix this on the frontend side as well so you wouldn't need custom server logic for this:
https://github.com/zeit/next.js/pull/2002

**Comment by janbaer** [comment]
That's indeed a big issue. Since Next.js is provided as a provided as a public framework for hosting server rendered React pages in an easy way, it should be possible also to host the solution on a subdomain. We decided to use Next.js for our next **admin-webapp**, but this will be hosted under a subdomain like www.some-domain.com/admin and this would actually not possible or am I wrong?

**Comment by nndung179** [comment]
@janbaer it actually possible. I have met the same problem like you. I used the solution above to solve. But I believe in the future it will be solved

**Comment by wesbos** [comment]
If anyone else is trying to run next on a sub-route with express, I had to use this express route to re-write the HMR pings:

```
app.use('/_next/*', (req, res) => {
  const newURL = req.originalUrl.replace('_next', 'admin/_next');
  res.redirect(newURL);
});
```

**Comment by timneutkens** [comment]
Thanks for that solution @wesbos 👌 Pretty sure it'll help people 👍

**Comment by janbaer** [comment]
@nndung179 @wesbos Thanks for your solutions, I'll try it. But I think it should be a feature of the framework, like the feature with for the CDN support.

**Comment by janbaer** [comment]
@wesbos - How have you changed to route that is listening for the pages:

I changed it in this way, but it will only return Ok but never the index page

```
  router.get(`/admin/*`, async ctx => {
    await handle(ctx.req, ctx.res);
    ctx.respond = false;
  });
```

Before line 1 was

```
router.get(`*`, async ctx => {
```

I'm using Koa with the koa-router.

**Comment by janbaer** [comment]
@wesbos I assume that it's because of this line in server/index.js

      '/:path*': async (req, res, params, parsedUrl) => {
        const { pathname, query } = parsedUrl
        await this.render(req, res, pathname, query)
      }

so, **admin/** will be not match to any file in pages but I would expect that I'm getting 404 but not 200 back... how have you solved the problem. I know that I can define my own routes, but that's not what I want to do because I decided for Next.js to have not so much to configure and with automatic routing.

**Comment by janbaer** [comment]
I tried to use the server side solution from @wesbos and it worked for me locally. But in case it would run on a server with Nginx as proxy-server which is only the /admin requests proxy to my Next.js page it wouldn't work, since the request to /_next wouldn't arrive in my server.js.

**Comment by manolkalinov** [comment]
This is actually a make or break issue if Next.js is to be used in a real production environment. @tscanlin's fix is only two lines long. Is there a way I/we can help get his fix in right now or at least formulate a temporary workaround?

**Comment by tscanlin** [comment]
@manolkalinov glad to hear my fix is working for you! Could you comment on the PR I opened to help make a case for getting this change landed? I'll add a comment too. https://github.com/zeit/next.js/pull/2002

**Comment by wesbos** [comment]
Finding issues with my method - the webpack HMR breaks and they live-reload takes 1-2 seconds instead of being instant. I guess because you can't proxy the eventsouce through express.

**Comment by vanmik** [comment]
I'm also trying to make next.js app work from subfolder. I add `appPrefix` in config, and now all `_next/*` links in <script> tags are correct. But all this links are 404, files still are available only from root. How to fix it?

**Comment by wesbos** [comment]
I figured out a (poor and non-permanant) fix for my hot reloading not working.

https://github.com/zeit/next.js/blob/f4345e79c254cff935fa5c4a7949ee4b0834cd78/client/webpack-hot-middleware-client.js#L1

I edited this file to include my new path. 

It would be ideal if with path would take the assetPrefix into account as well, since everything else does.

**Comment by gcpantazis** [comment]
#2186 fixes the `_next` half of this. `static` is something that I'd like to add configuration for as well, but I'd likely wait on that until after #2186 is landed.

**Comment by nndung179** [comment]
Currently, I solved it by extending **NextScript** & **Head** component of next.js with my own component. Allow passing props (prefix) to component that I defined. It worked very well!

**Comment by wesbos** [comment]
@nndung179 Can you share your extended component?

**Comment by nndung179** [comment]
Hi @wesbos !!!

Here is full demo:
[https://github.com/nndung179/nextjs-full-demo](https://github.com/nndung179/nextjs-full-demo)

In this demo will contain:
+ Babel config
+ PostCSS config
+ Webpack config: to add plugins allow develop CSS using SCSS
+ Redux
+ Material UI
+ Custom Document
+ Custom Server
+ Next Routes
+ Data fetching
+ Gulp
+ Dynamic import

The idea to configure `prefix` in domain is I created 2 components `Head` & `NextScript` and allow to pass props called `prefix`. Then in `_document.js` instead of using:

```js
import Document, { Main, Head, NextScript } from 'next/document'
```
Now
```js
import Document, { Main } from 'next/document'

// custom component
import NextScript from '@/components/NextJS/NextScript';
import Head from '@/components/NextJS/Head';
```

**Comment by wesbos** [comment]
Thanks a lot - will take a look. 

I keep hitting lots of HMR issues and 404s in my app. Are there any plans to make core handle sub-dir installations anytime soon?

**Comment by jthegedus** [comment]
+1 Regarding "Are there any plans to make core handle sub-dir installations anytime soon?"

**Comment by dvakatsiienko** [comment]
+1 we also need support for supplying assets ( `static`, `_app`) `prefix` configuration to make possible to serve multiple `next` apps on the same domain using `nginx` `reverse proxy` functionality.

For example:

`www.domain.com/app-one`

`www.domain.com/app-two`

**Comment by johhansantana** [comment]
so from all the stuff gathered here I think this is the easiest way to do so for the moment:

in your `next.config.js`
```
  assetPrefix: '/ofertap1/',
```
![image](https://user-images.githubusercontent.com/9933887/28727229-1d654fb2-7392-11e7-9b05-4d3e975dcab0.png)

notice my alias here will be `ofertap1`

and custom server
```javascript
const express = require('express');
const next = require('next');

const dev = process.env.NODE_ENV !== 'production';
const app = next({ dev });
const handle = app.getRequestHandler();

const prefix = '/ofertap1';
app.prepare()
.then(() => {
  const server = express();
  const router = express.Router();

  // use next routes
  server.use(`${prefix}`, router);
  server.use(`${prefix}/static`, express.static('static'));
  server.use(handle);

  router.get('*', (req, res) => {
    return handle(req, res)
  });

  server.listen(5001, (err) => {
    if (err) throw err;
    console.log(`> Ready on http://localhost:5001${prefix}`);
  });
});
```

then all your links to files/routes should have a prefix of `/ofertap1` (excluding imports)

you don't have to change your `page` folder it works with and without navigating with the prefix (not sure if everything will work without the prefix after doing all the steps though)

I'm sure there's a better way instead of having to change every link/route to add the prefix but for now this works for me if anyone ends up wanting a quick solution.

**Comment by rayhooker** [comment]
John, This is very close to working.  I have proxypass to my nodejs.  The initial page works and knows the prefix for all assets but once it tries hot reload it does not prefix it:
/_next/webpack-hmr 

It then gets in an unusable state and tries load from the wrong locations and the page crashes.

Any thoughts and what version of nextjs are you running?  I am on nextjs 2.4.7!  That appears to be the latest that npm installs.

**Comment by rayhooker** [comment]
Actually I figured out the problem. I was using the process.env.NODE_ENV value to set the prefix for links and css.  I needed to set it as an initial prop so that it would render the same value on the client if it rerenders for some reason.  

I had to proxypass though the /_next otherwise it gets errors trying to hot reload as the on-demand-entries-client does not know about the prefix so it goes to the original server so I forward /_next to /.

Hopefully they can fix that so it also uses assetPrefix or perhaps a different setting to say "my server is elsewhere".  Otherwise you can't have multiple nextjs apps serving the same url.

**Comment by tscanlin** [comment]
Thank you @johhansantana for sharing your script. That definitely helped me to get assetPrefix working for running multiple next.js apps on a single domain. Dev worked fine for me, but I had trouble with static exporting. So in the spirit of sharing I thought I'd post the script that I am now using to get static exports working, hopefully it helps someone. I'd love to hear ideas about any better solutions for this.

```js
const globby = require('globby')
const spawn = require('child_process').spawnSync
const config = require('../next.config.js')

const fs = require('fs-extra')

const prefix = config.assetPrefix

globby(`out/**/page`).then((files) => {
  const folder = files[0]
  const newFolder = folder.split('/page').join(prefix)
  console.log('Copying: ', folder, newFolder)
  fs.copySync(folder, newFolder)
  fs.moveSync(newFolder, folder + prefix)
  globby(folder + prefix + '/**/*.js').then((editFiles) => {
    console.log(editFiles);
    editFiles.forEach((file) => {
      fs.readFile(file, 'utf8', function (err,data) {
        if (err) {
          return console.log(err)
        }

        const newString = "window.__NEXT_REGISTER_PAGE('" + prefix
        const rootPathString = "window.__NEXT_REGISTER_PAGE('/'"

        let result = data
        if (data.indexOf(newString) === -1 && data.indexOf(rootPathString) === -1) {
          result = data.replace(/window.__NEXT_REGISTER_PAGE\(\'/g, newString)
        } else if (data.indexOf(rootPathString) > -1) {
          // If it's the index page then don't leave the trailing slash.
          result = data.replace(/window.__NEXT_REGISTER_PAGE\(\'\/\'/g, newString + "'")
        }

        fs.writeFile(file, result, 'utf8', function (err) {
           if (err) return console.log(err)
        })
      })
    })
  }).catch((e) => {
    console.log(e)
  })
  return files
}).catch((e) => {
  console.log(e)
})
```

**Comment by niccai** [comment]
+1

**Comment by rayhooker** [comment]
So I am close to John's approach working.  If I load directly from the browser command line it works, and any links to /index always work, but others do not. Note that I have app running with a prefix of /src.  Basically the link for /index works but I don't even see a call to the backend in the network tab of the developers in Chrome.  If I add the following:
       server.use(`${prefix}/_next`, (req, res) => {
            console.log("found it and fixing it")
            console.log("Old #1", req.url) 
            req.url = req.url.replace(/\/src/g, ""); 
            console.log("New #1", req.url)
            next()
        });
and  
   server.get(`${prefix}/_next`, (req, res) => {
        console.log("A next page call...")
        return app.render(req, res, '/_next', req.params)
    });

In this case I am able to view the pages other than / (or /index) but in all cases the transfer of the some files remain pending:

http://localhost:3000/src/_next/72b9db3e-ed16-443f-9e35-15d0702ae245/page/index (pending)
http://localhost:3000/src/_next/72b9db3e-ed16-443f-9e35-15d0702ae245/page/_error/index.js (pending) 
-- or--
http://localhost:3000/src/schools (succeeds)
http://localhost:3000/src/_next/72b9db3e-ed16-443f-9e35-15d0702ae245/page/schools (Pending)
http://localhost:3000/src/_next/72b9db3e-ed16-443f-9e35-15d0702ae245/page/_error/index.js (Pending)

Clearly it is not forwarding/ handling the _next HMR calls I guess.  Thoughts? ... note that the full code is pasted below.

co(function * () {
    yield app.prepare()
    console.log(`Connecting to ${MONGO_URL}`)
    const db = yield MongoClient.connect(MONGO_URL) 
    const server = express()
    server.use(body.json())
        server.use((req, res, next) => {
        console.log("Use body.json",req.url)
        req.db = db
        next()
    })
    // server.set('baseURL', '/src')
    server.use(`${prefix}`,express.Router())
    if (process.env.NODE_ENV === 'production') {
        prefix = '/src'
        server.use(`${prefix}/static`, express.static('static'));

        server.use(`${prefix}/_next`, (req, res) => {
            console.log("found it and fixing it")
            console.log("Old #1", req.url) 
            req.url = req.url.replace(/\/src/g, ""); 
            console.log("New #1", req.url)
            next()
        });
    }
    const wrapAsync = handler => (req, res) => handler(req, res)
    .then(result => res.json(result))
    .catch(error => res.status(500).json({ error: error.message }));

    server.get('/api/school', wrapAsync(async function(req, result) {
        var find_params = {};
        var school = {}; 
        if (req.query.unit_code !== undefined) {
            find_params.unit_code = req.query.unit_code; 
            if (req.query.year !== undefined) {
                find_params.year = req.query.year;
            }
            school.profile  = await db.collection('profile').find(find_params).toArray();
            school.proficiencies  = await db.collection('proficiency').find(find_params).toArray();
            school.teachers  = await db.collection('teachers').find(find_params).toArray();
            school.environment  = await db.collection('environment').find(find_params).toArray();
            return school
        } else {
            return result.status(404).json({err: "Need to specify unit_code"});
        };
    }))

    server.get('/api/schools', wrapAsync(async function(req, result) { 
            return  await db.collection('profile').find({}).toArray();
    }))

    server.get(`${prefix}/schools`, (req, res) => {
        return app.render(req, res, '/schools', req.params)
    });

    server.get(`${prefix}/index`, (req, res) => {
        return app.render(req, res, '/index', req.params)
    });

    server.get(`${prefix}/_next`, (req, res) => {
        console.log("A next page call...")
        return app.render(req, res, '/_next', req.params)
    });

    server.get(`${prefix}/districts`, (req, res) => {
        return app.render(req, res, '/districts', req.params)
    });

    server.get(`${prefix}/school`, (req, res) => {
        const { pathname, query } = parse(req.url, true)
        console.log("School params")
        console.log(Object.assign(params, query))
        return app.render(req, res, '/school', Object.assign(params, query))
    });

    server.get(`${prefix}`, (req, res) => {
        return app.render(req, res, '/index', req.params)
    });
 
    server.get('*', (req, res) => { 
        req.db = db
        console.log("Get ****", req.url)

        const { pathname, query } = parse(req.url, true)

        const params = match(pathname)
        console.log("Test url")
        if (params === false) {
            console.log("Replace src on get old ", req.url)
            req.url = req.url.replace(/\/src/g, "");
            console.log("Replace src on get new ", req.url)
        handle(req, res)
        return
        }
        console.log("Passed")
        console.log(Object.assign(params, query))
        // if (!dev) {
        //     console.log("Replace src on get old ", req.url)
        //     req.url = req.url.replace(/\/src/g, "");
        //     console.log("Replace src on get new ", req.url)
        // }
        app.render(req, res, '/school', Object.assign(params, query))
        next()
    })
    server.listen(PORT)
    console.log(`Listening on ${PORT}`)

**Comment by callumlocke** [comment]
For anyone using Next 3's new dynamic `import()` support, you might encounter the problem that 'chunks' do not honour your custom `assetPrefix`.

For example, if you do `import('../data').then(...)` somewhere in a clientside module, that will cause webpack to make a 'chunk' that's separate from your page/app/vendor bundles, and downloaded as something like `/_next/webpack/chunks/---data-acbeb0d2-27e8-4f2c-9772-b307327d8992.js`. (The `assetPrefix` setting is applied correctly to page bundles and stuff like app/commons, but not to chunks.)

Here's my config to get everything working under `/foo`:

```js
module.exports = {
  assetPrefix: '/foo', // affects page bundles and other assets
  webpack: (config) => {
    config.output.publicPath = `/foo${config.output.publicPath}`; // affects 'chunks'
    return config;
  },
};
```

This still leaves two other kinds of URL unprefixed – but fortunately these are dev-only features (and prefixing is probably only necessary in production for most people):

- `/_next/webpack-hmr`
- `/_next/on-demand-entries-ping?page=/some/page`

I've made a specific bug report for this assetPrefix issue at #2759.

**Comment by aviramga** [comment]
@johhansantana @davidnguyen179 Im trying to follow your solutions with nextjs export method. But for some reason I can't seem to configure NGINX correctly. The page is loaded so I know that the rules are valid but all imports that start with "/prefix-path/" doesn't seem to be redirected to the correct location. Any idea why?

**Comment by alexindigo** [comment]
This change might help you to solve the problem – https://github.com/zeit/next.js/pull/3332
In my case I just rename `_next` dir of the export into my custom directory and stick it on CDN.

**Comment by stoffern** [comment]
i solved it here: https://github.com/stoffern/extensive-react-server 

i find nextJs a good option but it has some core setup that dont work well. as you can not edit the setup completely for all your needs. Extensive React Server can..

- add as many apps as you want, 
- static folders, 
- files etc.. 
- add custom route files or folders with files
- authentication on files and react apps
- webpack setup fully editable with functions

**Comment by cheshirecode** [comment]
I'm trying to split up a single Nextjs app into multiple apps on the same subdomain with Nginx and thanks to https://github.com/zeit/next.js/issues/257#issuecomment-277729383 and https://github.com/zeit/next.js/issues/257#issuecomment-321797741, have gotten all requests to `/_next/` covered by now.

1) Assume `PATH_PREFIX` is the prefix constant we need to add to have our app served at `/foo` subdomain.
2) Inject this script into `_document.js` before `<Main/>`. This prefixes `/_next/webpack-hmr` for HMR EventStream and all `/_next/on-demand-entries-ping?page=` HMR pings:
```
(function(isDebug) {
  var __windowEventSourceConstructor__ = window.EventSource;
  var __prefixHardcodedNextRequests = function(url) {
    return url.match('/_next/')
      ? url
          .replace('/_next/', '${PATH_PREFIX}/_next/')
          .replace('${PATH_PREFIX}${PATH_PREFIX}', '${PATH_PREFIX}')
      : url;
  };
  window.EventSource = function(url) {
    isDebug && console.log('eventstream', url);
    return new __windowEventSourceConstructor__(
      __prefixHardcodedNextRequests(url)
    );
  };

  var __XMLHttpRequestOpen__ = XMLHttpRequest.prototype.open;
  // turned out that we need to override Webpack's output.publicPath for chunks referenced in browser
  XMLHttpRequest.prototype.open = function() {
    isDebug && console.log('open', arguments);
    //arguments[1] = __prefixHardcodedNextRequests(arguments[1]);
    return __XMLHttpRequestOpen__.apply(this, arguments);
  };

  var __fetch__ = fetch;

  fetch = function() {
    isDebug && console.log('fetch', arguments);
    arguments[0] = __prefixHardcodedNextRequests(arguments[0]);
    return __fetch__.apply(this, arguments);
  };
})();
```

3) Modify Next's Webpack config to prefix Webpack chunks requested from browser like `/_next/2dab534d-29e0-4ff7-bf1f-6d3868836051/webpack/0a0998c3e67208d1416e.hot-update.json` AND `/_next/2dab534d-29e0-4ff7-bf1f-6d3868836051/webpack/5.067204aeccbc56133800.hot-update.js`
```
  webpack: (config, { dev }) => ({
    ...config,
    output: {
      ...config.output,
      publicPath: `${dev ? `${PATH_PREFIX}` : ''}${config.output.publicPath}`
    }
  }),
```

4) Now Nginx is able to route all requests including HMR-related starting with `/foo/` to one app, `/bar/` to another. Prod was very easy, dev with hot-reloading took ages to figure out but thanks to all the discussion, we should have some stopgap measures while waiting for v5.

**Comment by danistefanovic** [comment]
I've encountered the same problem with HMR but in a slightly different setup (multi-tenant environment with dynamic subdomains):

* example.com -> /
* example.com/about -> /about
* tenant-a.example.com -> /_sub/:tenant/
* tenant-a.example.com/:postId -> /_sub/:tenant/:postId

In my case, the HMR wasn't working under the subdomains because `http://tenant-a.example.com/_next/webpack-hmr` wasn't available. Adding a proxy to redirect to the "main" domain solved the issue:

```js
// server.js
const express = require('express');
const next = require('next');
const proxy = require('http-proxy-middleware');

// ...

const dev = process.env.NODE_ENV !== 'production';
const app = next({ dev });
const handle = app.getRequestHandler();

app.prepare().then(() => {
  const server = express();

  // ↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓

  if (dev) {
    // Make HMR work
    server.use(
      '/_next',
      proxy((pathname, req) => Boolean(req.subdomains.length), {
        target: 'http://localhost:3000',
        changeOrigin: true,
      }),
    );
  }

  // ↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑

  // ...

  server.get('*', (req, res) => handle(req, res));

  server.listen(port, (err) => {
    if (err) throw err;
    console.log(`> Ready on http://localhost:${port}`);
  });
});

```

Maybe this helps someone.

**Comment by sergiodxa** [comment]
The team is working right now on a feature for Next.js that will let you **run multiple applications on the same domain**. The **feature is called [multi-zone](https://github.com/zeit/next.js/tree/multi-zones)**, and @arunoda crafted an example called [with-zones](https://github.com/zeit/next.js/tree/multi-zones/examples/with-zones) to check how it works, [it's working here](https://with-zones.now.sh/).

Note that this is still on **pre-alpha**, but if any of you want to use it and give feedback or report any issue will be welcomed.

**Comment by cheshirecode** [comment]
Cool stuff! Still, this feature solve a different problem and ties itself to *micro-proxy*. I'll extract out a working example when possible to illustrate.

**Comment by sergiodxa** [comment]
You don't need to use micro-proxy to use this feature, you just need a proxy in front of your two apps, that's usual to have two different server side apps running on the same domain.

**Comment by atuttle** [comment]
Glad to hear that a fix is coming. Very much looking forward to it. Just so that I can plan accordingly for an upcoming product launch, should we expect multi-zone to go stable in the near future? How near?

**Comment by overkazaf** [comment]
@sergiodxa Great to hear that. I was suffering with not being able to deploy two next apps in my own single domain, if you've already complete this feature, please let me know ASAP, thanks!

**Comment by tomaswitek** [comment]
Hi @overkazaf it looks like @arunoda from Zeit team is working on it. I just got an answer in Slack. Maybe you could try the `multi-zones` branch in meantime?
I think it is a very important feature for a lot of people, if you need help guys, we are here 😏

**Comment by tomaswitek** [comment]
Hi @arunoda I just tried out your `multi-zones` branch locally at it works perfectly. Great work!
One question, when I click `blog` link than a normal server side request happens and the whole page is reloaded. Will it stay like that or do you guys plan to support also client side routing between apps without reload? Thx!

**Comment by glenswift** [comment]
Thanks for your work and this zones stuff you have done. One quick question: do you have any estimates when this will become production ready? We're thinking about migrating to next.js but hosting several apps on the same domain is crucial for us, so it would be great to be aligned with your roadmap.

Thanks.

**Comment by stoffern** [comment]
@glenswift I would recommend using my solution then. see more here:
https://github.com/velop-io/server

You can serve as many reactRoutes you want and add authentication to those you want..

**Comment by timneutkens** [comment]
@glenswift will be released *very soon* 👍

**Comment by tomaswitek** [comment]
@stoffern but your solution is a replacement, you can't use it together with next.js right?

**Comment by stoffern** [comment]
It is a definition on what you need. 
NextJS was made for hosting simple apps with config first.
Velop server is code first, and is made for hosting multiple apps, multiple custom routes, multiple different authentication modules for every routes ++
Read more here: https://github.com/velop-io/server#why-not-nextjs

simple example would be
```
var server = new Server({
  environment: 'development',
  hostname: 'localhost',
  port: 3000,
});

route1 = server.addReactRoute('path to app')
route2 = server.addReactRoute('path to app', '/prefix')
route2.addAuthentication(passportfunction)

server.start()
```

**Comment by zenflow** [comment]
The multi-zones branch is 404ing 😞 https://github.com/zeit/next.js/tree/multi-zones

**Comment by zenflow** [comment]
@timneutkens It looks like the multi-zones branch was merged, since there's now a [Multi Zones section in the README](https://github.com/zeit/next.js/#multi-zones)? Is this issue resolved?

**Comment by tomaswitek** [comment]
Hi @zenflow, yep it was merged and [released in 5.0.0](https://zeit.co/blog/next5#zones).
So yes, this issue can be probably closed 😏

**Comment by timneutkens** [comment]
Yep this should be closed 🎉

**Comment by joshuaquek** [comment]
I used:

In my `routes.js` file:
```javascript
// ---- API Routes ----
const express = require('express')
const router = express.Router()

// -------- NextJs API Routes --------
router.get(`*`,  (req, res, next) => {
  return app.getRequestHandler()  // where app is nextjs app object
})

module.exports = router
```

In my `server.js` file:
```javascript
const server = require('express')()
let routes = require('routes.js')
server.use('/my_custom_base_url', routes)

server.listen(port, (error) => { // Start server to listen on specified port
  if (error) throw error
  console.log('Server started.')  
})
```

and in my `next.config.js` file:
```javascript
module.exports = {
  assetPrefix: '/my_custom_base_url'
  publicRuntimeConfig: { 
    staticFolder: `/my_custom_base_url/static`
  },
  webpack: (config) => {
    // Fixes npm packages that depend on `fs` module
    config.node = {
      fs: 'empty'
    }
    return config
  }
}

```

This fixed all of my routing issues and will make the base url of the entire nextjs webapp to have a base url of '/my_custom_base_url' 

My version of NextJs is `next@7.0.2`

Hope this helps!

I have also mentioned this as a solution to Issue #2901

**Comment by bgarrett-rt** [comment]
For anyone being redirected to this issue ticket trying to just get next mounting to a subdirectory, here's what finally ended up working for me.

```
// next.config.js
module.exports = {
   ...
 // With this line, next will try to load routes like `/my_custom_base_url/_next/static/development/pages/_app.js` but they will 404 because next is expecting /_next/ at the root of the url.
  assetPrefix: '/my_custom_base_url',
  ...
}
```

// We can use middleware to rewrite the url so next handles it as if it wasn't in a subdirectory i.e. `/my_custom_base_url/_next/static/development/pages/_app.js` =>  `/_next/static/development/pages/_app.js`

```
// server.js - using express.js - app is instance of `next({...})`. Note: Due to scope next() is not the same var nor type as next() in the 3rd argument of the middleware below
app.prepare().then(() => {
  const server = express();
  
  // The rewire middleware - not. req.url is mutable but req.originalUrl isn't, or at least has no effect if mutated
  server.use(function(req, res, next) {
     req.url = req.originalUrl.replace('my_custom_base_url/_next', '_next');
     next(); // be sure to let the next middleware handle the modified request. 
  });

  server.get('*', (req, res) => {
    return handle(req, res);
  });
  ... 
```
I have not thoroughly tested yet, but this worked when deployed as well. For my use case, I have only one next.js app and it is merely accessible via /my_custom_base_url. Due to service routing, requests above /my_custom_base_url/ (i.e. /_next/) are not intercepted at all by the next service. Hope this helps save someone else an afternoon of  head scratching.

**Comment by hect1c** [comment]
Hey,

I'm currently facing the same problem where I have a basePath and need to have the site working on http://domain/basePath

I've tried both solutions above but it doesn't seem to be working. Without the basePath it works fine, but with it I still get a 404 page not found. Any ideas what I could be missing? 

Below is my server.js

`const next = require('next')
const server = require('express')()

const port = parseInt(process.env.PORT, 10) || 3000
const dev = process.env.NODE_ENV !== 'production'
const app = next({ dev })
const handle = app.getRequestHandler()

app.prepare().then(() => {
    server.use((req,res, next) => {
        req.url = req.originalUrl.replace('teachonline/_next', '_next');
        next(); // be sure to let the next middleware handle the modified request.
    });

    server.get('*', (req, res) => {
        return handle(req,res);
    });

    server.listen(port, err => {
        if (err) throw err
        console.log(`>> Ready on http://localhost:${port}`)
    });
});`

**Comment by godmar** [comment]
First time experience with next.js, was disappointed that building + copying to personal web directory didn't work out of the box.  For anyone googling this in the future, I'll leave a note to hopefully save them the hour I spent trying to google this. To get a quick "see what next.js means by deployment" view, you must create a `next.config.js` file like this as pointed out [earlier in this thread](https://github.com/zeit/next.js/issues/257#issuecomment-445509191):
```
$ cat next.config.js 
module.exports = {
  distDir: 'build',
  assetPrefix: '/~yourpid/out'
};
```
then you can make the [setup](https://github.com/zeit/next.js/#setup) example and run
```
npx next build && npx next export
/bin/cp -R out ~/public_html/out
```
which will allow you to see something at `shared.institutional.edu/~yourpid/out` assuming your sys admins have [standard userdir](https://serverfault.com/questions/326810/replicating-userdir-apache-functionality-in-nginx) functionality set up in your institution's Apache or nginx. 

This doesn't solve the larger issue(s) raised in this thread, but at least will allow a newcomer to see what next.js calls a "production build."  Note that unlike [CRA](https://facebook.github.io/create-react-app/) next.js requires both a build and an export step; 'build' here doesn't build anything that could be deployed. The [main documentation](https://github.com/zeit/next.js/#static-html-export) doesn't mention this step directly, but the [extra step of setting assetPrefix](https://github.com/zeit/next.js/wiki/Deploying-a-Next.js-app-into-GitHub-Pages#extra-step-for-project-page-sites) is hidden in a page titled [Deploying a Next.js app into GitHub Pages](https://github.com/zeit/next.js/wiki/Deploying-a-Next.js-app-into-GitHub-Pages#extra-step-for-project-page-sites) linked from there for people wanting to deploy on github pages. I would be kind of nice if the default setup didn't assume domain root access and used relative paths only, but perhaps my I'm stuck in the 1990s and in any event would apply only to 100% static applications which next.js perhaps doesn't target.

**Comment by tanobi92** [comment]
> If anyone else is trying to run next on a sub-route with express, I had to use this express route to re-write the HMR pings:
> 
> ```
> app.use('/_next/*', (req, res) => {
>   const newURL = req.originalUrl.replace('_next', 'admin/_next');
>   res.redirect(newURL);
> });
> ```

When I deploy to server, then the states in redux be lost after change page. How to fix it?

**Comment by alxhghs** [comment]
Had a hard time figuring this out initially. Wrote a blog post on the solution https://www.alexhughes.dev/blog/multiple-next-apps/

**Comment by kepeterson** [comment]
> Hey I have a similiar situation with basePath, but can't get out of it. I have tried my own solution at first, but when it didn't work 100% I used @bgarrett-rt 's code for the server.
> 
> `server.js`
> 
> ```js
> const express = require('express');
> const compression = require('compression');
> const next = require('next');
> const { config } = require('./env-config');
> 
> const dev = config['NODE_ENV'] !== 'production';
> const PORT = config['PORT'] || (dev ? 4100 : 80);
> const ROOT_PATH = config['ROOT_PATH'] || '/';
> const app = next({ dev });
> const handle = app.getRequestHandler();
> 
> app.prepare().then(() => {
>     const server = express();
> 
>     server.use(compression());
>     server.use((req, res, next) => {
>         req.url = dev ? req.originalUrl :
>             req.originalUrl.replace(ROOT_PATH, '');
> 
>         next();
>     })
>     server.get('*', (req, res) =>
>         handle(req, res));
> 
>     server.listen(PORT, err => {
>         if (err) throw err;
>         //eslint-disable-next-line
>         console.log(`> Ready on http://localhost:${PORT}`);
>     });
> });
> ```
> 
> `next.config.js`
> 
> ```js
> const withPlugins = require('next-compose-plugins');
> const sass = require('@zeit/next-sass');
> const css = require('@zeit/next-css');
> const fonts = require('next-fonts');
> const images = require('next-images');
> const { config } = require('./env-config');
> 
> const dev = config['NODE_ENV'] !== 'production';
> const rootPath = dev ? '' : config['ROOT_PATH'];
> 
> module.exports = withPlugins([
>     [sass], [css], [fonts], [images]
> ], {
>     publicRuntimeConfig: {
>         rootPath
>     },
>     assetPrefix: rootPath
> });
> ```
> 
> The initial load works fine - no 404's whatsoever, but after clicking any link it just loads the page infinitely. While watching the requests, all the needed chunks are loaded succesfully, it just looks like if after loading of the chunks something goes wrong and the page isn't being rendered.
> 
> I guess that it is also worth noting that i'm using the `rootPath` config option in my Link Wrapper which just adds the rootPath to href. They look ok, just as I need: `http://domain.com/basePath/pagesDir/pagesDirComponent`.
> 
> Can someone point me in some kind of direction? I'm banging my head at this for half a day now...

Were you able to make any progress on this? I seem to be experiencing the exact same issue. I'm re-writing my urls the same way you are so I can use a `ROOT_PATH`. Nothing 404s on initial load but when I try to use a `<Link/>` nothing happens. If I navigate to the route from the address bar, everything is fine.

**Comment by vinodloha** [comment]
Hi all, whoever is facing this problem this below solution has worked absolutely spot-on for me in a non Now environment. This guides you to achieve Zones like capability that Zeit Now provides. 

https://levelup.gitconnected.com/deploy-your-nextjs-application-on-a-different-base-path-i-e-not-root-1c4d210cce8a

**Comment by kepeterson** [comment]
> Hi all, whoever is facing this problem this below solution has worked absolutely spot-on for me in a non Now environment. This guides you to achieve Zones like capability that Zeit Now provides.
> 
> https://levelup.gitconnected.com/deploy-your-nextjs-application-on-a-different-base-path-i-e-not-root-1c4d210cce8a
> 
> One improvement i did, was to create a routes.js file which just returns an object of routes, so that i don't have to create wrapper components to prepend base path prefix.

Does hot-loading work in development mode? And do Link components work?

**Comment by vinodloha** [comment]
> > Hi all, whoever is facing this problem this below solution has worked absolutely spot-on for me in a non Now environment. This guides you to achieve Zones like capability that Zeit Now provides.
> > https://levelup.gitconnected.com/deploy-your-nextjs-application-on-a-different-base-path-i-e-not-root-1c4d210cce8a
> > One improvement i did, was to create a routes.js file which just returns an object of routes, so that i don't have to create wrapper components to prepend base path prefix.
> 
> Does hot-loading work in development mode? And do Link components work?

Yes, hot reloading works as usual. As far as Link component is concerned, you should write a wrapper that adds base_path to all your links. This blog I referenced has example for it.

**Comment by Naxos84** [comment]
> 
> 
> Hi all, whoever is facing this problem this below solution has worked absolutely spot-on for me in a non Now environment. This guides you to achieve Zones like capability that Zeit Now provides.
> 
> https://levelup.gitconnected.com/deploy-your-nextjs-application-on-a-different-base-path-i-e-not-root-1c4d210cce8a
> 
> One improvement i did, was to create a routes.js file which just returns an object of routes, so that i don't have to create wrapper components to prepend base path prefix.

How does that work with dynamic routes?
like "/posts/1234" as "/posts/[id]"

**Comment by Timer** [comment]
Next.js 9.5 was just released with support for a [Customizable Base Path](https://nextjs.org/blog/next-9-5#customizable-base-path), allowing you to mount multiple Next.js apps on one domain, so long as their subpaths are unique!

**Comment by marlonmleite** [comment]
@Timer if you want to change only the prefix/path of the static files?
My app run on `/` and the assets and files come from `/any _next` for example.
It's possible?

**Comment by arihantverma** [comment]
@Timer, as the original request says, we have multiple next apps running on the same domain. When we use base path and use proxy passes to handle, the one problem that we aren't able to mitigate is to keep links 'clean'. Say our basepath is `app1`. All our problems with asset prefix, api route path prefix, images path prefix is resolved, except that our links show `app1/about`. Is there a way to leverage `as` `Link` prop with basepath so that we can keep our anchor links the same?

**Comment by arihantverma** [comment]
We ended up using [`next-build-id`](https://github.com/nexdrew/next-build-id#readme) to [configure a custom build id](https://nextjs.org/docs/api-reference/next.config.js/configuring-the-build-id) where we gave a `app-<name>-<commitSHA>` signature and added a re write rule in our load balancer for incoming request host name having `app-<name>`

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
